### PR TITLE
feat: support any type

### PR DESCRIPTION
**Pull request by jannyHou**
Implements https://github.com/strongloop/loopback-next/issues/4255, which supports the `any` as the type for arbitrary model property.

*This is the first PR created to support `any` type. We need another PR to append the `any` type's openapi schema in `components.schemas`*

### Description 

Define the property in model as:
  
```ts
import {AnyType} from '@loopback/repository';
@property({
    // `@loopback/repository-json-schema` recognizes it
    type: 'any'
  })
  // specify `AnyType` for it
  arbitraryProp: AnyType
```

Generated json schema would be

```ts
arbitraryProp: {
    $ref: '#/definitions/AnyType'
}
```

Generated openapi spec would be

```ts
"components": {
    "schemas": {
      "Todo": {
        "title": "Todo",
        "properties": {
          "id": {
            "type": "number"
          },
          // ...other properties
          // The schema will be generated as a reference
          // TODO: to completely support this feature, we can provide swagger AnyType schema
          // out-of-the-box
          // see link https://swagger.io/docs/specification/data-models/data-types/#any
          "arbitraryProp": {
            "$ref": "#/components/schemas/AnyType"
          }
        },
        "required": [
          "title"
        ],
        "additionalProperties": false
      },
  },
}
```

-->

## Checklist

👉 [Read and sign the CLA (Contributor License Agreement)](https://cla.strongloop.com/agreements/strongloop/loopback-next) 👈

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [ ] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

👉 [Check out how to submit a PR](https://loopback.io/doc/en/lb4/submitting_a_pr.html) 👈

### CODE REVIEW

**Comment by jannyHou** [comment]
---

DISCUSSION POINT (the todo from last comment)

I plan to create anther PR adding the openapi shema for `any`.
According to https://swagger.io/docs/specification/data-models/data-types/#any, the openapi schema spec for any type will be:

```ts
components: {
   schemas: {
      AnyValue: {}
   }
}
```

The reason why I plan to integrate this schema into the rest server's spec is:

When a model has a property in type of another model, we usually expect user to define that model properly so that its schema is appended in the server OpenAPI spec's `components.schemas`. But the situation for property in `AnyType` is tricky, because it's a LoopBack type we defined in [@loopback/repository](https://github.com/strongloop/loopback-next/blob/master/packages/repository/src/types/any.ts#L13).
Therefore users don't have a model for it in their project's `/models` folder, like how they have `/models/address.model.ts` defined for a `Customer` model that contains an address property.

This PR supports the any type at json schema level. And due to the above reason, we need another PR which adds schema to the rest server's spec. And also update the todo example to have an arbitrary property.

WDYT?

**Comment by agnes512** [comment]
Thanks Janny for bring up the discussion.  I have two questions/concerns as I am not that familiar with this area:
1. Since LoopBack need the defined `AnyType` in the first place, what would it affect if we change it to an arbitrary type? 
2. What are we going to do with the `AnyType`?

**Comment by raymondfeng** [comment]
I'm not sure if OpenAPI spec json schema supports `any` type.

**Comment by raymondfeng** [comment]
Maybe `any` in json schema is `oneOf` the following:

- string | number | boolean | null 
- json array 
- json object

**Comment by raymondfeng** [comment]
You can use the following schema definition for any:

```js
properties:
{
  anyProp: true, // or anyProp: {}
}
```

File: .vscode/launch.json [hunk 1]
```diff
@@ -43,6 +49,15 @@
...
       "port": 5858
+    },
+    {
+      "type": "node",
+      "request": "launch",
+      "name": "Example",
+      "program": "${workspaceFolder}/examples/todo/index.js",
+      "outFiles": [
+        "${workspaceFolder}/examples/todo/**/*.js"
+      ]
```
Comment (raymondfeng) [inline .vscode/launch.json]: Please revert this change.

File: examples/todo/src/models/todo.model.ts [hunk 2]
```diff
@@ -41,6 +41,12 @@ export class Todo extends Entity {
   })
   remindAtGeo?: string; // latitude,longitude
 
+  @property({
+    type: 'any',
+  })
+  // eslint-disable-next-line @typescript-eslint/no-explicit-any
+  arbitraryProp?: any;
```
Comment (raymondfeng) [inline examples/todo/src/models/todo.model.ts]: Let's have a more meaningful name such as `tag`?

File: packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts [hunk 3]
```diff
@@ -179,6 +180,23 @@ describe('build-schema', () => {
         expectValidJsonSchema(jsonSchema);
       });
 
+      it('properly converts any properties', () => {
+        @model()
+        class TestModel {
+          @property({
+            type: 'any',
+          })
+          anyProp: AnyType;
```
Comment (raymondfeng) [inline packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts]: The TypeScript typing should not be `AnyType`. Use `any` or `unknown` instead.

File: packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts [hunk 4]
```diff
@@ -179,6 +180,23 @@ describe('build-schema', () => {
...
+        }
+
+        const jsonSchema = modelToJsonSchema(TestModel);
+        expect(jsonSchema.properties).to.deepEqual({
+          anyProp: {},
```
Comment (raymondfeng) [inline packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts]: Do we prefer `{}` or `true`?

File: packages/repository-json-schema/src/build-schema.ts [hunk 5]
```diff
@@ -193,6 +194,10 @@ export function stringTypeToWrapper(type: string | Function): Function {
       wrapper = Buffer;
       break;
     }
+    case 'any': {
+      wrapper = AnyType;
```
Comment (raymondfeng) [inline packages/repository-json-schema/src/build-schema.ts]: The wrapper is probably `Object`. `AnyType` is used internally for conversions.

File: packages/rest/src/__tests__/unit/ajv.service.unit.ts [hunk 6]
```diff
@@ -39,6 +39,53 @@ describe('Ajv service', () => {
...
+    }
+
+    function testAnyTypeWith(value: string) {
+      it(`with value ${value}`, async () => {
+        const ajv = new AjvProvider().value();
+        const validator = ajv.compile({
+          type: 'object',
+          properties: {
+            name: {type: 'string'},
+            arbitraryProp: {},
```
Comment (raymondfeng) [inline packages/rest/src/__tests__/unit/ajv.service.unit.ts]: Can we also test `arbitraryProp: true`?

File: packages/rest/src/__tests__/unit/ajv.service.unit.ts [hunk 7]
```diff
@@ -39,6 +39,53 @@ describe('Ajv service', () => {
     expect(result).to.be.true();
   });
 
+  // possible values for type any
+  const TEST_VALUES = {
+    string: 'abc',
+    number: 123,
+    object: {random: 'random'},
+    array: [1, 2, 3],
```
Comment (raymondfeng) [inline packages/rest/src/__tests__/unit/ajv.service.unit.ts]: Add `null`?

File: examples/todo/src/models/todo.model.ts (hunk 2 above)
Comment (jannyHou) [reply examples/todo/src/models/todo.model.ts]: `tag` sounds good

File: packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts (hunk 3 above)
Comment (jannyHou) [reply packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts]: good catch, updated to `any`

File: examples/todo/src/__tests__/acceptance/todo.acceptance.ts [hunk 8]
```diff
@@ -68,6 +68,21 @@ describe('TodoApplication', () => {
     expect(result).to.containDeep(todo);
   });
 
+  it('creates a todo with arbitrary property', async function() {
+    // Set timeout to 30 seconds as `post /todos` triggers geocode look up
+    // over the internet and it takes more than 2 seconds
+    // eslint-disable-next-line no-invalid-this
+    this.timeout(30000);
```
Comment (nabdelgadir) [inline examples/todo/src/__tests__/acceptance/todo.acceptance.ts]: Since there's no geocode in this test, you don't need this part

File: .vscode/launch.json [hunk 9]
```diff
@@ -24,7 +27,10 @@
       "request": "launch",
       "name": "Debug Current Test File",
       "program": "${workspaceRoot}/bin/mocha-current-file",
-      "runtimeArgs": ["-r", "${workspaceRoot}/packages/build/node_modules/source-map-support/register"],
+      "runtimeArgs": [
+        "-r",
+        "${workspaceRoot}/packages/build/node_modules/source-map-support/register"
+      ],
```
Comment (raymondfeng) [inline .vscode/launch.json]: I would rather revert this file as it's not related to this feature.

File: packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts (hunk 3 above)
Comment (raymondfeng) [reply packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts]: Not yet. I'm still seeing `AnyType`.

File: packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts (hunk 4 above)
Comment (jannyHou) [reply packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts]: @raymondfeng The reason I eventually chose `{}` is because the object can take in other properties like `description`, see [code](https://github.com/strongloop/loopback-next/blob/master/packages/repository-json-schema/src/build-schema.ts#L246-L254)
I verified `{description: 'some desc'}` is still a valid syntax for any type.
WDYT? I can switch to `true` if we want to align with the latest standard.

File: packages/rest/src/__tests__/unit/ajv.service.unit.ts (hunk 6 above)
Comment (jannyHou) [reply packages/rest/src/__tests__/unit/ajv.service.unit.ts]: cc @raymondfeng see the comment in https://github.com/strongloop/loopback-next/pull/4930#discussion_r398970633

File: packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts (hunk 4 above)
Comment (raymondfeng) [reply packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts]: As long as we test `true`, I'm good either way.

Comment (raymondfeng) [reply packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts]: See https://json-schema.org/draft/2019-09/json-schema-core.html#rfc.section.4.3.2

**Comment by raymondfeng** [comment]
Pls fix code lint error - https://travis-ci.com/github/strongloop/loopback-next/jobs/305196649

**Comment by jannyHou** [comment]
@agnes512 Thank you for reviewing and good questions:

> Since LoopBack need the defined AnyType in the first place, what would it affect if we change it to an arbitrary type?

`AnyType` is only for internal use(didn't realize that in the original implementation), it's replaced by type `any` now. For "what would it affect if we change it to an arbitrary type?" I am not sure what do you mean by "arbitrary type" here...since `any` already allows the arbitrary type :p

e.g. given a property `tag` that allows any value:

```ts
@property({type: 'any'})
tag?: any
```
Which part do you want to change and see how it affects?

> What are we going to do with the AnyType?

Since `AnyType` is not used anymore, I assume your question is about `any` now. 

- `any` is the type of your model property, like the `tag` property above
- the {type: 'any'} as the meta data of a model property, is consumed by the `@loopback/repository-json-schema` module, to generate the correct json schema (`{}` or `true`, still discuss with @raymondfeng ) for the `tag` field.
- eventually a property with type `any` will have `{}` (or `true`) as its openapi schema

File: packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts (hunk 4 above)
Comment (jannyHou) [reply packages/repository-json-schema/src/__tests__/integration/build-schema.integration.ts]: sounds good, I added another 2 set for `true`.

File: examples/todo/src/__tests__/acceptance/todo.acceptance.ts [hunk 10]
```diff
@@ -65,6 +65,14 @@ describe('TodoApplication', () => {
     expect(result).to.containDeep(todo);
   });
 
+  it('creates a todo with arbitrary property', async function () {
```
Comment (raymondfeng) [inline examples/todo/src/__tests__/acceptance/todo.acceptance.ts]: Nitpick: Let's use `async () =>` to be consistent.
